### ISSUE TITLE
Failed to automatically apply fixes suggested by rustc

### ISSUE THREAD

**Issue by asg0451** [rust-lang/rust#83988]
I ran this command: 
```
cargo +nightly clippy -Zunstable-options --fix --allow-dirty
```

and it told me to file a bug ;). Output below as requested:
```
$ cargo +nightly clippy -Zunstable-options --fix --allow-dirty
    Checking idx v0.1.0 (/home/miles/rust/lfm-idxr/idx)
warning: failed to automatically apply fixes suggested by rustc to crate `idx`

after fixes were automatically applied the compiler reported errors within these files:

  * idx/src/ingest_event.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0308]: mismatched types
  --> idx/src/ingest_event.rs:56:9
   |
56 |         self.profile_type.as_ref().map(|profile_type| (&self.channel, profile_type))
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `str`, found struct `std::string::String`
   |
   = note: expected enum `std::option::Option<(&str, &str)>`
              found enum `std::option::Option<(&std::string::String, &std::string::String)>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Original diagnostics will follow.

warning: this `if` has identical blocks
  --> idx/src/ingest_event.rs:43:71
   |
43 |           } else if channel == "instagram" && profile_type == "comment" {
   |  _______________________________________________________________________^
44 | |             Some(fb_ts_to_epoch(profile.get("created_time")?.as_str()?).ok()?)
45 | |         } else {
   | |_________^
   |
   = note: `#[warn(clippy::if_same_then_else)]` on by default
note: same as this
  --> idx/src/ingest_event.rs:41:63
   |
41 |           if channel == "facebook" && profile_type == "comment" {
   |  _______________________________________________________________^
42 | |             Some(fb_ts_to_epoch(profile.get("created_time")?.as_str()?).ok()?)
43 | |         } else if channel == "instagram" && profile_type == "comment" {
   | |_________^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#if_same_then_else

warning: manual implementation of `Option::map`
  --> idx/src/ingest_event.rs:56:9
   |
56 | /         if let Some(profile_type) = self.profile_type.as_ref() {
57 | |             Some((&self.channel, profile_type))
58 | |         } else {
59 | |             None
60 | |         }
   | |_________^ help: try this: `self.profile_type.as_ref().map(|profile_type| (&self.channel, profile_type))`
   |
   = note: `#[warn(clippy::manual_map)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_map

warning: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
  --> idx/src/tok_norm.rs:63:1
   |
63 | fn split_token_at_hm<'a>(tok: &'a str) -> Vec<String> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::needless_lifetimes)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes

warning: this `if` has identical blocks
  --> idx/src/ingest_event.rs:43:71
   |
43 |           } else if channel == "instagram" && profile_type == "comment" {
   |  _______________________________________________________________________^
44 | |             Some(fb_ts_to_epoch(profile.get("created_time")?.as_str()?).ok()?)
45 | |         } else {
   | |_________^
   |
   = note: `#[warn(clippy::if_same_then_else)]` on by default
note: same as this
  --> idx/src/ingest_event.rs:41:63
   |
41 |           if channel == "facebook" && profile_type == "comment" {
   |  _______________________________________________________________^
42 | |             Some(fb_ts_to_epoch(profile.get("created_time")?.as_str()?).ok()?)
43 | |         } else if channel == "instagram" && profile_type == "comment" {
   | |_________^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#if_same_then_else

warning: length comparison to zero
  --> idx/src/tok_norm.rs:64:13
   |
64 |     assert!(tok.len() > 0);
   |             ^^^^^^^^^^^^^ help: using `!is_empty` is clearer and more explicit: `!tok.is_empty()`
   |
   = note: `#[warn(clippy::len_zero)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_zero

warning: length comparison to zero
  --> idx/src/tok_norm.rs:83:8
   |
83 |     if token.len() == 0 {
   |        ^^^^^^^^^^^^^^^^ help: using `is_empty` is clearer and more explicit: `token.is_empty()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_zero

warning: manual implementation of `Option::map`
  --> idx/src/ingest_event.rs:56:9
   |
56 | /         if let Some(profile_type) = self.profile_type.as_ref() {
57 | |             Some((&self.channel, profile_type))
58 | |         } else {
59 | |             None
60 | |         }
   | |_________^ help: try this: `self.profile_type.as_ref().map(|profile_type| (&self.channel, profile_type))`
   |
   = note: `#[warn(clippy::manual_map)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_map

warning: length comparison to zero
   --> idx/src/tok_norm.rs:104:13
    |
104 |     assert!(token.len() > 0);
    |             ^^^^^^^^^^^^^^^ help: using `!is_empty` is clearer and more explicit: `!token.is_empty()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_zero

warning: explicit lifetimes given in parameter types where they could be elided (or replaced with `'_` if needed by type declaration)
  --> idx/src/tok_norm.rs:63:1
   |
63 | fn split_token_at_hm<'a>(tok: &'a str) -> Vec<String> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(clippy::needless_lifetimes)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes

warning: length comparison to zero
  --> idx/src/tok_norm.rs:64:13
   |
64 |     assert!(tok.len() > 0);
   |             ^^^^^^^^^^^^^ help: using `!is_empty` is clearer and more explicit: `!tok.is_empty()`
   |
   = note: `#[warn(clippy::len_zero)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_zero

warning: length comparison to zero
  --> idx/src/tok_norm.rs:83:8
   |
83 |     if token.len() == 0 {
   |        ^^^^^^^^^^^^^^^^ help: using `is_empty` is clearer and more explicit: `token.is_empty()`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_zero

warning: length comparison to zero
   --> idx/src/tok_norm.rs:104:13
    |
104 |     assert!(token.len() > 0);
    |             ^^^^^^^^^^^^^^^ help: using `!is_empty` is clearer and more explicit: `!token.is_empty()`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_zero

warning: 6 warnings emitted

warning: 6 warnings emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.80s
```

rustc nightly version: rustc 1.52.0-nightly (51748a8fc 2021-03-05)
Ubuntu 20.04

The code might be confidential... but I might be able to provide it on request.
I am in a project within a cargo workspace, if that helps. I typically use stable, but use nightly for this command.

Let me know if you have any questions!

**Comment by asg0451** [comment]
adding return type to the fn in question fixed it:
```
pub fn channel_profile_type(&self) -> Option<(&str, &str)> {
        self.profile_type
            .as_ref()
            .map(|profile_type|  { (&self.channel, profile_type) })
    }

```
->
```
pub fn channel_profile_type(&self) -> Option<(&str, &str)> {
        self.profile_type
            .as_ref()
            .map(|profile_type| -> (&str, &str) { (&self.channel, profile_type) })
    }
```

why did that fix it?

**Comment by SkiFire13** [comment]
Minimal reproducible example:

```rust
pub fn test(s: Option<&String>) -> Option<&str> {
    match s.as_ref() {
        Some(s) => Some(s),
        None => None,
    }
}
```

> why did that fix it?

Previously the closure return type was inferred, which prevented deref coercion from happening. By specifying the return type you removed the type inference step and allowed deref coercion to occur. See the [nomicon](https://doc.rust-lang.org/nomicon/coercions.html) for more details.

**Comment by asg0451** [comment]
Interesting. i dont quite follow, but that's fine. Nevertheless, clippy asked me to file this issue, so I did ;)

**Comment by SkiFire13** [comment]
Should have been fixed in https://github.com/rust-lang/rust-clippy/pull/7085

**Comment by AloeareV** [comment]
> Should have been fixed in [rust-lang/rust-clippy#7085](https://github.com/rust-lang/rust-clippy/pull/7085)

Was this fixed in this PR?

**Comment by SkiFire13** [comment]
From what I can see yes, at least the minimal example is fixed and the fix seems to address the fundamental problem (coercions happening before but not after). If you find cases where this is still wrong feel free to report them.

**Comment by fmease** [comment]
So from reading the discussion, this seems to have been fixed? Closing as fixed then. Please ping me if misjudged and I will of course reopen this issue.

In the future, please report Clippy-related issues over at https://github.com/rust-lang/rust-clippy.

Closing as fixed.
